### PR TITLE
Fresh-install UAT repairs across single-user, multi-user, SQLite and PostgreSQL

### PR DESCRIPTION
**Pull request by rmusser01**
## Summary

Fresh-install UAT exposed failures in authentication, saved chat identity and retries, image attachments, media and study flows, and World Book behavior across single-user and multi-user setups on SQLite and PostgreSQL. This PR fixes those boundaries using the existing authentication, database and shared UI patterns, with regression coverage and a running issue tracker.

The final review fixes prevent concurrent session-key deletion, require a durable scoped write before acknowledging offline Notes saves, recover Chat titles safely across account changes, upgrade existing PostgreSQL source catalogs, and preserve grounded legacy Chatbook flashcard saves. Shared transport, translations and thin Next page boundaries address the remaining frontend feedback. All 14 Qodo review threads are resolved, and the refreshed review reports zero open findings. The alleged Flashcards syntax error was disproved with the compiler and dismissed by the reviewer.

The branch has been rebased onto freshly fetched dev. Generated Playwright captures are excluded from the PR and preserved locally. The human requester authorized merging this repair PR before resuming UAT, starting with UAT261's original TestBot exact-output check. The next full fresh-install matrix has not started.

## Change summary

This PR fixes issues uncovered during fresh-install UAT across single-user and multi-user setups on SQLite and PostgreSQL. It repairs authentication, chat persistence and retries, image attachments, media and study workflows, and World Book behavior—including cross-user data isolation and PostgreSQL connection cleanup. The fixes follow existing authentication and database patterns, add regression coverage, and maintain a running issue tracker while keeping generated Playwright captures out of the PR.

## Validation

- Manual credential repair: 113 focused and adjacent auth/storage/bootstrap tests pass. Final real browser lifecycle: 3 passes, 0 skips, 0 retries on Node 20.20.2, covering device, legacy and session reload/reopen behavior. Independent source review is clear.
- Qodo backend repairs: 70 combined controls pass, followed by 35 final focused controls after refinements, all without skips. Actual PostgreSQL checks cover old-schema upgrade/data retention, repeated initialization, concurrent readers and connection/transaction cleanup. Legacy save tests cover ownership, linkage, public excerpts and rejection before writes.
- Root frontend review fixes: 46 focused tests pass (9 Settings, 32 reading-progress, 5 Help), plus 12 static route-title and 7 Login/navigation tests. Independent source review is clear. All 18 active Help catalogues and 21 extension mirrors contain the new messages.
- Notes and title repairs: 18 focused UI tests and 18 existing web title tests pass. Three causal title-provenance cases also pass after the final test-fixture lint correction, with zero title-source/test lint diagnostics. Independent source review is clear.
- Earlier integrated repairs retain 110 Chat Retry controls, 50 World Book consumer controls, 111 secondary operation/parser controls, Notes/doc/UI verification, and hosted accessibility evidence. These focused results do not certify completed final UAT.
- Canonical OpenAPI regeneration and fresh fingerprint check pass in the CI-matching Python environment. The delta is exactly two field descriptions; 2097 paths and 3207 schemas are unchanged. Frontend types regenerated successfully.
- Scoped ESLint has no new errors. Nine touched backend production files have zero Bandit findings; remaining Ruff diagnostics match the baseline. Bandit does not analyze TypeScript.

Final CI also exposed three older Billing response fixtures missing their JSON Content-Type. The reviewed header-only correction passes all 42 Settings cookie/logout and form lifecycle tests with unchanged assertions and zero scoped lint diagnostics.

Historical failures and unsuccessful intermediate attempts remain recorded. Final required hosted CI is a merge gate.

## Tracking and remaining acceptance

[Running UAT tracker](https://github.com/rmusser01/tldw_server/blob/35e46b7a146d5dbd1cdcd2fc659ba135abc99d45/Docs/Reviews/FRESH_INSTALL_SINGLE_MULTI_UAT_TRACKER_2026_09_14.md) · [Qodo dispositions](https://github.com/rmusser01/tldw_server/blob/35e46b7a146d5dbd1cdcd2fc659ba135abc99d45/Docs/Reviews/PR2967_QODO_REVIEW_2026_09_18.md) · [Integration verification](https://github.com/rmusser01/tldw_server/blob/35e46b7a146d5dbd1cdcd2fc659ba135abc99d45/Docs/Reviews/PR2967_INTEGRATION_VERIFICATION_2026_09_18.md)

282 UAT findings: 281 verified and UAT261 open, with the 14 Qodo findings separately documented in the review ledger. UAT261 remains the first post-merge check with its exact criterion unchanged; then resume the fresh SQLite/PostgreSQL × single/multi-user workflow matrix, including actual image attachment.

## Migration and rollback

Existing PostgreSQL World Books without recorded owners are preserved but hidden until trusted maintenance assignment. Follow the [ownership upgrade guide](https://github.com/rmusser01/tldw_server/blob/35e46b7a146d5dbd1cdcd2fc659ba135abc99d45/Docs/Deployment/Database/world-book-owner-upgrade.md); never assign old books to the first requester. Legacy Chatbook card creation accepts linked public question/answer turns; arbitrary, unlinked or structured content still fails before writes. Back up populated databases before deployment; reverting application code alone is not a database rollback.

### CODE REVIEW

File: apps/packages/ui/src/components/Flashcards/tabs/ManageTab.tsx [hunk 1]
```diff
@@ -211,8 +211,7 @@ const ManageExpertFilters: React.FC<ManageExpertFiltersProps> = ({
       <Popover
         trigger="click"
         placement="bottomLeft"
-        content={
-          <div className="w-72 space-y-2">
+        content=<div className="w-72 space-y-2">
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Flashcards/tabs/ManageTab.tsx:214]: Verified against the actual ManageTab.tsx with TypeScript 5.9.3: transpileModule reports zero diagnostics and emits the module successfully. TypeScript explicitly defines JsxAttributeValue as StringLiteral | JsxExpression | JsxElement | JsxSelfClosingElement | JsxFragment, so JSX elements are valid direct attribute initializers here. The existing expressions are intentional valid TSX; no syntax repair is required. This disposition is recorded in Docs/Reviews/PR2967_QODO_REVIEW_2026_09_18.md.

File: apps/packages/ui/src/components/Notes/hooks/useNotesEditorState.tsx [hunk 2]
```diff
@@ -173,24 +232,32 @@ export function useNotesEditorState(deps: UseNotesEditorStateDeps) {
   const [openingLinkedChat, setOpeningLinkedChat] = React.useState(false)
 
   // ---- offline draft state ----
-  const [offlineDraftQueue, setOfflineDraftQueue] = React.useState<Record<string, OfflineDraftEntry>>({})
-  const [offlineDraftQueueHydrated, setOfflineDraftQueueHydrated] = React.useState(false)
+  const [offlineDraftQueue, setOfflineDraftQueue] = useNotesAuthorityState<Record<string, OfflineDraftEntry>>(authorityScope, {})
+  const [offlineDraftQueueHydrated, setOfflineDraftQueueHydrated] = useNotesAuthorityState(authorityScope, false)
+  const offlineDraftStorageKey = authorityScope ? `${NOTES_OFFLINE_DRAFT_QUEUE_STORAGE_KEY}:${authorityScope}` : null
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Notes/hooks/useNotesEditorState.tsx:237]: Fixed in a55e86167f. Offline save now requires a confirmed authority, hydrated owner-scoped queue, synchronous storage write and matching readback before acknowledging success or clearing dirty state. Missing authority, storage failure and mismatched readback preserve unsaved edits and show localized recovery text. Nine authority-race and six offline drafting/sync controls pass; all 18 active catalogues and 21 extension mirrors contain the recovery message. Independent review is clear.

File: tldw_Server_API/app/core/Ingestion_Sources/service.py [hunk 3]
```diff
@@ -205,12 +379,13 @@ async def ensure_ingestion_sources_schema(db) -> None:
...
-    await _ensure_sqlite_column(
-        db,
-        table_name="ingestion_source_items",
-        column_name="present_in_source",
-        column_sql="INTEGER NOT NULL DEFAULT 1",
-    )
+    if db.backend == "sqlite":
+        await _ensure_sqlite_column(
+            db,
+            table_name="ingestion_source_items",
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Ingestion_Sources/service.py:386]: Fixed in 3420b66228. The database-management migration adds present_in_source for existing PostgreSQL tables, preserving legacy rows with a default of true. It checks pg_catalog for the actual search-path-resolved table before issuing ALTER, avoiding an exclusive table lock on normal requests. Four tests using the official PostgreSQL fixture pass without skips, covering a real pre-column table, read/upsert/filter behavior, repeated initialization, retained bindings and an active concurrent reader.

File: tldw_Server_API/app/api/v1/schemas/chat_knowledge_schemas.py [hunk 4]
```diff
@@ -44,6 +59,10 @@ def _normalize_tags(cls, value: Optional[list[str]]) -> Optional[list[str]]:
 
     @model_validator(mode="after")
     def _validate_scope(self) -> "KnowledgeSaveRequest":
+        if not self.snippet:
+            raise ValueError("Snippet must contain visible answer text")
+        if self.make_flashcard and (not self.flashcard_front or not self.flashcard_back):
+            raise ValueError("A flashcard requires both a question and an answer")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/api/v1/schemas/chat_knowledge_schemas.py]: Fixed in 3420b66228. Verified the consumer request model at tldw_chatbook b0dadf19414f5f8faf69b854d8e59007d275083e: it omits the two card fields. When both fields are absent, the server now derives a card only from an owned linked assistant answer and its owned user parent in the same conversation; the requested excerpt must occur in the public assistant answer. Explicit front/back remains supported. All 20 knowledge-save tests pass, including workspace/global scope and foreign-parent controls. Partial explicit fields, unlinked content, hidden reasoning, ungrounded excerpts and structured JSON still return 422 before either write; arbitrary snippets cannot safely produce a meaningful card without additional generation. The compatibility limitation is documented in the PR.

File: apps/packages/ui/src/components/Common/PageHelpModalHost.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  if (!isOpen) return null
+  if (HelpModal) return <HelpModal />
+
+  return (
+    <div className="fixed bottom-4 right-4 z-50 max-w-sm rounded border border-border bg-surface p-4 shadow-lg" role={failed ? "alert" : "status"}>
+      <p className="text-sm text-text">
+        {failed
+          ? t("help.loadFailed", "Help could not be loaded. Reconnect, save any edits, then reload the page.")
+          : t("help.loading", "Loading help…")}
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Common/PageHelpModalHost.tsx:59]: Fixed in 9344f3ae95. Added translated help.loadFailed, help.loading, help.confirmReload and help.reload messages to all 18 active locale catalogues and 21 Chrome-format mirrors, including aliases. All 39 JSON files and required keys validate; the five existing Help host tests pass.

File: apps/packages/ui/src/components/Option/Settings/tldw.tsx [hunk 6]
```diff
@@ -94,18 +98,52 @@ export const TldwSettings = () => {
...
+
+  useEffect(() => {
+    setBillingServerUrl(null)
+    if (authMode !== 'multi-user' || !isLoggedIn || !serverUrl || isQuickstartWebUiSameOriginServerUrl(serverUrl)) return
+    const controller = new AbortController()
+    const timeout = setTimeout(() => controller.abort(), 5000)
+    let cancelled = false
+    void (async () => {
+      try {
+        const response = await fetch(`${serverUrl.replace(/\/$/, '')}/openapi.json`, { signal: controller.signal })
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Option/Settings/tldw.tsx]: Fixed in 9344f3ae95. Billing discovery uses bgRequest with the configured absolute root /openapi.json URL, noAuth, a 5-second timeout, cancellation and suppressed capability-probe outage reporting. It preserves the four required billing GET paths and rejects late responses after logout. The transport regression fails against the former direct fetch; the final nine Settings form/lifecycle cases pass with unchanged timeouts and assertions.

File: tldw_Server_API/app/api/v1/API_Deps/ChaCha_Notes_DB_Deps.py [hunk 7]
```diff
@@ -441,24 +443,30 @@ def _apply_sqlite_tuning(db_instance: CharactersRAGDB) -> None:
...
-            use_wal=False,
-            synchronous=None,
-            foreign_keys=True,
-            busy_timeout_ms=1000,
-            temp_store=None,
-        )
-        conn.execute("SELECT 1")
+        if db_instance.backend_type == BackendType.POSTGRESQL:
+            with db_instance.transaction() as transaction:
+                transaction.execute("SELECT 1")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/api/v1/API_Deps/ChaCha_Notes_DB_Deps.py]: Fixed in 3420b66228. Probe SQL and backend handling now live in core/DB_Management/chacha/health.py. The dependency delegates while retaining its independent operation boundary, sanitized logging and boolean result. Existing actual PostgreSQL transaction/cleanup and failed-probe controls, plus SQLite controls, pass in the combined 70-test backend run.

File: apps/packages/ui/src/components/Option/Playground/PlaygroundForm.tsx [hunk 8]
```diff
@@ -1604,6 +1612,62 @@ export const PlaygroundForm = ({
...
+      const matchesActiveChat = selection?.kind === "character"
+        ? activeChat.serverChatAssistantKind === "character" &&
+          String(activeChat.serverChatCharacterId) === selection.id
+        : selection?.kind === "persona"
+          ? activeChat.serverChatAssistantKind === "persona" &&
+            String(activeChat.serverChatAssistantId) === selection.id
+          : activeChat.serverChatMetaLoaded &&
+            activeChat.serverChatAssistantKind == null &&
+            activeChat.serverChatAssistantId == null &&
+            activeChat.serverChatCharacterId == null;
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Option/Playground/PlaygroundForm.tsx]: Fixed in 9344f3ae95. Each of the three guards explicitly checks === null or === undefined, preserving the existing saved-assistant metadata semantics. Scoped ESLint reports zero errors and no warning on the changed lines.

File: apps/packages/ui/src/components/Review/__tests__/ViewMediaPage.permalink.test.tsx [hunk 9]
```diff
@@ -367,6 +438,96 @@ describe('ViewMediaPage Stage 3 permalinks', () => {
...
+    mocks.detailById['1'] = { media_id: 1, source: { title: 'Aster' }, content: { text: 'Original Aster source' } }
+    let release!: (value: unknown) => void
+    mocks.getNavigationResume.mockReturnValue(new Promise(resolve => { release = resolve }))
+    renderMediaPage('/media?id=1')
+    await waitFor(() => expect(screen.getByTestId('reading-scroller').scrollTop).toBe(160))
+    await waitFor(() => expect(mocks.getNavigationResume).toHaveBeenCalled())
+
+    await act(async () => {
+      release(remembered ? { media_id: '1', node_id: 'chapter-1', title: 'Opening', level: 1 } : null)
+      await new Promise(resolve => setTimeout(resolve, 1100))
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/components/Review/__tests__/ViewMediaPage.permalink.test.tsx]: Fixed in 9344f3ae95. The three debounce checks now advance fake setTimeout/clearTimeout timers inside act after render readiness, and restore real timers after each test. The actual scroll, chapter-jump and persisted-position assertions remain intact. All 32 permalink tests pass.

File: apps/tldw-frontend/pages/characters.tsx [hunk 10]
```diff
@@ -1,3 +1,8 @@
 import dynamic from "next/dynamic"
+import Head from "next/head"
 
-export default dynamic(() => import("@/routes/option-characters"), { ssr: false })
+const Characters = dynamic(() => import("@/routes/option-characters"), { ssr: false })
+
+export default function CharactersPage() {
+  return <><Head><title>Characters | tldw</title></Head><Characters /></>
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/tldw-frontend/pages/characters.tsx]: Fixed in 9344f3ae95. A web-only withPageTitle adapter owns Next Head and the 12 pages delegate through it to their browser-only route components. Static titles remain available to SSR. All 12 route-title cases and two metadata-governance checks pass. Next imports remain outside shared extension UI.

File: apps/tldw-frontend/pages/chat/index.tsx [hunk 11]
```diff
@@ -1,3 +1,17 @@
-import dynamic from "next/dynamic"
+import dynamic from 'next/dynamic';
+import Head from 'next/head';
+import { useActiveChatTitle } from '@/hooks/useActiveChatTitle';
 
-export default dynamic(() => import("@/routes/option-chat"), { ssr: false })
+const ChatRoute = dynamic(() => import('@/routes/option-chat'), { ssr: false });
+
+export default function ChatPage() {
+  const { title } = useActiveChatTitle();
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/tldw-frontend/pages/chat/index.tsx]: Fixed in 9344f3ae95. The Chat page now only re-exports the web route adapter, which owns the title hook, Head and SSR fallback. The shared extension route remains free of Next-specific imports. All 18 existing web title integration, SSR and late-persistence cases pass after the separately reviewed title-authority repair.

File: apps/tldw-frontend/pages/login.tsx [hunk 12]
```diff
@@ -136,12 +137,24 @@ const LoginPage = () => {
       try {
         const { tldwClient } = await import("@/services/tldw/TldwApiClient")
         const config = await tldwClient.getConfig()
+        let multiUser = config?.authMode === "multi-user"
+        if (config?.serverUrl && !multiUser && !config.apiKey && config.authSource !== "cookie-session") {
+          try {
+            const metadata = await tldwClient.getFirstRunMetadata()
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/tldw-frontend/pages/login.tsx]: Fixed in 9344f3ae95. The Login page now only delegates to a web route adapter; existing auth-mode discovery, configuration updates and Next navigation behavior moved intact behind that boundary. All seven existing Login/navigation tests pass.

File: apps/packages/ui/src/hooks/useActiveChatTitle.ts [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+      config: authorityLoading ? null : config,
+      controller: new AbortController(),
+      saving: false,
+    };
+    active.current = owner;
+  } else if (!authorityLoading && config) {
+    if (!owner.config) owner.config = config;
+    else if (!connectionAuthoritiesMatch(config, owner.config)) owner.controller.abort();
+  }
+  const ready = !authorityLoading && !!config && !owner.controller.signal.aborted && !temporaryChat;
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/packages/ui/src/hooks/useActiveChatTitle.ts]: Fixed in a55e86167f. An aborted or changed authority gets a fresh title owner. For the same selected server conversation, title reads and renames remain blocked until that owner observes its own metadata reload; previous-owner loading evidence cannot release the fence. Genuinely new local selections recover normally. Independent review found and closed both the new-local-selection and successive-authority gaps. Three causal provenance tests and all 18 existing web title tests pass, with zero lint diagnostics in the final title source/test.

File: tldw_Server_API/app/api/v1/endpoints/flashcards.py [hunk 14]
```diff
@@ -191,11 +198,13 @@ def _build_flashcard_verification_units(cards: list[dict[str, Any]]) -> list[Art
...
         if back:
+            answer_claim = f"Question: {front}\nAnswer: {back}" if front else back
             units.append(
                 ArtifactVerificationUnit(
                     unit_id=f"flashcard:{index}:back",
-                    text=back,
-                    claims=[back],
+                    text=answer_claim,
+                    claims=[answer_claim],
+                    metadata={"requires_semantic_verification": True},
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/api/v1/endpoints/flashcards.py]: Fixed in 3420b66228. Flashcard claim construction moved unchanged into core/Flashcards/verification.py and the endpoint delegates to it. The question/answer unit and requires_semantic_verification metadata are preserved. Seven grounding controls and four endpoint/mocked-provider verification controls pass.

**Comment by rmusser01** [comment]
All 14 review findings have been addressed individually, with fix commits and verification evidence in their threads. Qodo reports zero open findings and all 14 threads are resolved. The final candidate is `4752bbe5aa07d7439deef60050ab2c0aaa9a1ffa`; the last commit records review closure and does not change application or test source.

The source fixes cover session-credential concurrency, durable offline Notes saves, title ownership recovery, existing PostgreSQL source upgrades, safe legacy Chatbook compatibility, and the requested transport/localization/architecture improvements. [The review ledger](https://github.com/rmusser01/tldw_server/blob/4752bbe5aa07d7439deef60050ab2c0aaa9a1ffa/Docs/Reviews/PR2967_QODO_REVIEW_2026_09_18.md) includes every disposition and retained validation limit.

The requester's Change summary remains verbatim in the PR description. Final required CI is running. This PR is the accumulated UAT repair candidate; the original TestBot UAT261 remains open and is first after the requested merge, before the next full SQLite/PostgreSQL single/multi-user matrix. Generated Playwright captures remain excluded from the PR.

@qodo-code-review Please review the final candidate for any remaining issues.

**Comment by rmusser01** [comment]
Final CI found three Billing test-fixture failures on 4752bbe5aa. They reproduce locally and arise because the synthetic JSON responses advertise text/plain, while the real shared transport uses Content-Type to select parsing. The actual backend returns application/json.

Commit `35e46b7a146d5dbd1cdcd2fc659ba135abc99d45` adds that header to seven fixtures only. All 42 Settings cookie/logout and form lifecycle tests pass; assertions, timeouts, cancellation and real transport remain unchanged. Independent review is clear and scoped ESLint reports zero diagnostics. The failed hosted and local results remain in the review ledger. Required CI is running on this updated candidate.
